remotes/workqueue.go

```diff
@@ -0,0 +1,83 @@
+package remotes
+
+import (
+	"context"
+
+	"golang.org/x/sync/errgroup"
+)
+
+type workItem struct {
+	do   func(ctx context.Context) error
+	done chan struct{}
+	deps []chan struct{}
+}
+
+func newWorkItem(do func(ctx context.Context) error, deps ...chan struct{}) *workItem {
+	return &workItem{
+		do:   do,
+		done: make(chan struct{}),
+		deps: deps,
+	}
+}
+
+func (wi *workItem) process(ctx context.Context) error {
+	defer close(wi.done)
+	for _, d := range wi.deps {
+		select {
+		case <-ctx.Done():
+			return nil
```

Comment:
You are supposed to return `ctx.Err()` I think?

Reply:
Not completely sure, I just want to skip the work item and drain the queue (it is just an early return in case of global cancellation, the workitem in itself has not failed). I wonder what the common pattern is in this case.

Reply:
Also, please note that the workQueue here is not exposed and has not been designed as a completely general purpose solution (I designed it for the problem at hand: bounded parallelism and task dependencies). I would not bother to much about the shortcomings of its design if it fits the particular problem it has to solve.

Reply:
This error return here becomes the return value for the goroutine in the `errgroup` and eventually becomes the result of `wq.workGroup.Wait()` → `waitForCompletion()` → `fixupImage()` → `FixupImage`. If there's another path which results in `FixupImage` returning non-success when this happens then I didn't find it.

If the context is cancelled (or times out) then the overall call really should return the ErrCancelled/ErrTimedout (didn't check the names) you'll get from `ctx.Err()` not success, so it can be propagated, otherwise you'll end up with the caller thinking everything went fine when it was actually cancelled (or timeout, etc), I think that's a problem for the problem at hand, not just the general case.

I'm pretty sure that
```go
case <-ctx.Done():
  return ctx.Err()
```
is the common pattern and the correct thing to do, you can see it in the docs on `Done` in the struct defintin at https://golang.org/pkg/context/#Context

Reply:
Thanks for the clarification, I will do that. I might need to do some error reporting tests though. I'll look into that :)